### PR TITLE
[native] expose module publishing to native context

### PR DESCRIPTION
**Pull request by davidiw**
In Aptos, we plan on allow modules to be published within the context of
the VM. This exposes thsi via a NativeContext so we can implement this
via native functions.

Note, I'm just playing around here. I can't fully grok how we'd go about testing this in an efficient fashion.

### CODE REVIEW

File: language/move-vm/runtime/src/loader.rs [hunk 1]
```diff
@@ -626,10 +626,10 @@ impl Loader {
     //
     // All modules in the bundle to be published must be loadable. This function performs all
     // verification steps to load these modules without actually loading them into the code cache.
-    pub(crate) fn verify_module_bundle_for_publication(
+    pub(crate) fn verify_module_bundle_for_publication<'a>(
         &self,
         modules: &[CompiledModule],
-        data_store: &mut impl DataStore,
+        data_store: &'a mut dyn DataStore,
```
Comment (wrwg) [inline language/move-vm/runtime/src/loader.rs:632]: Why do we need to switch to `dyn`?

**Review by wrwg — APPROVED** [review]

File: language/move-vm/runtime/src/native_functions.rs [hunk 2]
```diff
@@ -139,6 +139,34 @@ impl<'a, 'b> NativeContext<'a, 'b> {
         self.data_store.events()
     }
 
+    pub fn publish_modules(
```
[the remark this answers is not in the document]
Comment (tnowacki) [reply language/move-vm/runtime/src/native_functions.rs:142]: And I don't think there is much benefit (unless I'm missing something) to doing it this way, vs `NativeExtensions`, since you will not be able to access this module during this transaction anyway (as it won't be able to link ahead of time)

File: language/move-vm/runtime/src/loader.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (tnowacki) [reply language/move-vm/runtime/src/loader.rs:632]: (just like, running tests before and after and seeing the time change would be enough)
It probably don't be any different, just sort of curious :)

File: language/move-vm/runtime/src/module_validator.rs [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+    compatibility::Compatibility,
+    errors::{verification_error, Location, PartialVMError, PartialVMResult, VMResult},
+    normalized, CompiledModule, IndexKind,
+};
+use move_core_types::{account_address::AccountAddress, vm_status::StatusCode};
+use move_vm_types::data_store::DataStore;
+use std::collections::BTreeSet;
+use tracing::warn;
+
+pub(crate) fn validate_and_compile_modules<'a>(
```
[the remark this answers is not in the document]
Comment (davidiw) [reply language/move-vm/runtime/src/module_validator.rs:17]: recommendations appreciated!

File: language/move-vm/runtime/src/native_functions.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (davidiw) [reply language/move-vm/runtime/src/native_functions.rs:142]: Might have missed something in your proposal. The data store is internal to the move crate, so there's no way to expose this. The only way to currently publish a module is via the session interface, which isn't available in the context of an executing script(function). Though happy to identify a cleaner path forward?

[the remark this answers is not in the document]
Comment (wrwg) [reply language/move-vm/runtime/src/native_functions.rs:142]: The lazy publishing (like setting a mark in the native extension that some stuff has to been published after return from execution) is definitely the more cumbersome path.  We like to programmatically publish code, and errors resulting from validation should become execution errors. Why not do it like this PR suggests? It is the most natural way.

File: language/move-vm/runtime/src/loader.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (wrwg) [reply language/move-vm/runtime/src/loader.rs:632]: Well "sort of curious" creates work for other people. Do you actually expect an impact? Than please say so.

File: language/move-vm/runtime/src/module_validator.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (tnowacki) [reply language/move-vm/runtime/src/module_validator.rs:17]: I feel like `validate_bundle`, `validate_modules` or something might be enough. It isn't compiling anything, and got a bit nervous at first when I saw the name. Thought y'all were looking to run the compiler in the VM lol

File: language/move-vm/runtime/src/loader.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (tnowacki) [reply language/move-vm/runtime/src/loader.rs:632]: I get ya, not trying to make too much work. I was probably over thinking it. And have `dyn` on the mind from @vgao1996's recent PR

While not necessary, all I was indicating was running `cargo test -p <some vm crate>` (maybe `move-vm-transactional-tests`) and seeing if the time to run changed at all.
But I think its also fine to tell me to take a hike :p

File: language/move-vm/runtime/src/native_functions.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (tnowacki) [reply language/move-vm/runtime/src/native_functions.rs:142]: > The only way to currently publish a module is via the session interface, which isn't available in the context of an executing script(function)

But it is available in the session. You can do more than one thing in the session. (more on that below)

> Why not do it like this PR suggests? It is the most natural way.

I worry about threading more and more things into the natives like this. Particularly, I worry about long term implications for performance. For example, we'd like to probably in the future do a lot more in module loading for more efficient execution (we really don't take advantage of type information currently, for example). At the very least, Dario was very big on this direction back in the day. 
I think adding additional module loading in the middle of execution might complicate things, and isn't strictly necessary for this to work. 

I know there are a lot of might/maybe/sorta/somedays here. My point is simply keeping things in the runtime simple will make all sorts of endeavors easier in the future. 

I'm also not entirely convinced this is the "most natural way", or at the very least I would argue its subjective.

> Though happy to identify a cleaner path forward?

To put what I was hinting at more clearly, I would imagine something like 

- Native function adds the module/package bytes to the `NativeContextExtensions`
  - You can do verification in here too if you really want to deserialize and verify the module during runtime (though it is very expensive time wise).
  - Linking will definitely be harder here
- At the end of the transaction, grab the module bytes out of `NativeContextExtensions`
- Call `session.publish_module_bundle` on the bytes

TL;DR we can do this already with `NativeContextExtensions`. I'm not sure that this way presented in this PR is really better or worse in the short term. But in the long term, I worry it might complicate things.

File: language/move-vm/runtime/src/module_validator.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (tnowacki) [reply language/move-vm/runtime/src/module_validator.rs:17]: We aren't supposed to let me name things though, unless_you_like_really_long_over_explained_yet_somehow_vague_names ;)

File: language/move-vm/runtime/src/loader.rs (hunk 1 above)
Comment (vgao1996) [reply language/move-vm/runtime/src/loader.rs:632]: I'm not sure about this. The following playground example compiles without any problems:
```
trait Foo {
    fn foo(&self);
}

fn g(x: &dyn Foo) {
    
}

fn f(x: &impl Foo) {
    g(x)
}
```
You can definitely use an `&imp`l as `&dyn`, but not the other way around due to the Sized constrant

File: language/move-vm/runtime/src/loader.rs [hunk 4]
```diff
@@ -659,12 +659,12 @@ impl Loader {
     //   bundle. Basically, this represents the modules appears before `M` in the bundle vector.
     // - the `bundle_unverified` argument tracks the modules that have not been verified when `M`
     //   is being verified, i.e., the modules appears after `M` in the bundle vector.
-    fn verify_module_for_publication(
+    fn verify_module_for_publication<'a>(
```
Comment (vgao1996) [inline language/move-vm/runtime/src/loader.rs:662]: Is the explicit lifetime really necessary? I can't see why the compiler couldn't derive this automatically.

File: language/move-vm/runtime/src/native_functions.rs [hunk 5]
```diff
@@ -139,6 +139,34 @@ impl<'a, 'b> NativeContext<'a, 'b> {
...
+            sender,
+            self.data_store,
+            self.resolver.loader(),
+        )
+        .map_err(|e| e.to_partial())?;
+
+        // All modules verified, publish them to data cache
+        for (module, blob) in compiled_modules.into_iter().zip(modules.into_iter()) {
+            self.data_store
+                .publish_module(&module.self_id(), blob)
```
Comment (sblackshear) [inline language/move-vm/runtime/src/native_functions.rs:158]: I think this change should be considered with caution. The VM is very careful to maintain the [invariant](https://github.com/move-language/move/blob/main/language/documentation/spec/vm.md#code-cache) that code cache does not get updated in the middle of execution, and I'm not sure what might break if this invariant doesn't hold. I can't concretely point to things that will go wrong, but in the spirit of being conservative about invariants in safety-critical code I would suggest that we either: 
- use a different scheme like the one suggested by Todd above that preserves the invariant
- hold on this PR until someone further investigates the implications of breaking the invariant + updates the spec accordingly

**Comment by wrwg** [comment]
I'm looking at the alternative path to implement this, and will let the PR know once I see clearer.

**Comment by wrwg** [comment]
> 

Things are looking good. It is clean since once we are "finishing" a transaction, we can still execute arbitrary code (even a chained transaction). The effects and the abort behavior of both would then be combined. During finalization we also walk over all native extensions, thus can see if a "code publishing" extension has requests.
